src/codegen/unwinding.cpp

```diff
@@ -410,23 +417,83 @@ static unw_word_t getFunctionEnd(unw_word_t ip) {
     return pip.end_ip;
 }
 
+static bool inASTInterpreterExecuteInner(unw_word_t ip) {
+    static unw_word_t interpreter_instr_end = getFunctionEnd((unw_word_t)interpreter_instr_addr);
+    return ((unw_word_t)interpreter_instr_addr <= ip && ip < interpreter_instr_end);
+}
+
+static bool inGeneratorEntry(unw_word_t ip) {
+    static unw_word_t generator_entry_end = getFunctionEnd((unw_word_t)generatorEntry);
+    return ((unw_word_t)generatorEntry <= ip && ip < generator_entry_end);
+}
+
+
+static inline unw_word_t get_cursor_reg(unw_cursor_t* cursor, int reg) {
+    unw_word_t v;
+    unw_get_reg(cursor, reg, &v);
+    return v;
+}
+static inline unw_word_t get_cursor_ip(unw_cursor_t* cursor) {
+    return get_cursor_reg(cursor, UNW_REG_IP) - 1;
+}
+static inline unw_word_t get_cursor_bp(unw_cursor_t* cursor) {
+    return get_cursor_reg(cursor, UNW_TDEP_BP);
+}
+
+template <typename FrameFunc>
+bool unwindProcessFrame(unw_word_t ip, unw_word_t bp, unw_cursor_t* cursor, FrameFunc func) {
+    CompiledFunction* cf = getCFForAddress(ip);
+    bool jitted = cf != NULL;
+    if (!cf) {
+        if (inASTInterpreterExecuteInner(ip)) {
+            cf = getCFForInterpretedFrame((void*)bp);
+            assert(cf);
+        }
+    }
+
+    if (!cf)
+        return false;
+
+    PythonFrameIteratorImpl info(jitted ? PythonFrameId::COMPILED : PythonFrameId::INTERPRETED, ip, bp, cf);
+    if (jitted) {
+        // Try getting all the callee-save registers, and save the ones we were able to get.
+        // Some of them may be inaccessible, I think because they weren't defined by that
+        // stack frame, which can show up as a -UNW_EBADREG return code.
+        for (int i = 0; i < 16; i++) {
+            if (!assembler::Register::fromDwarf(i).isCalleeSave())
+                continue;
+            unw_word_t r;
+            int code = unw_get_reg(cursor, i, &r);
+            ASSERT(code == 0 || code == -UNW_EBADREG, "%d %d", code, i);
+            if (code == 0) {
+                info.regs[i] = r;
+                info.regs_valid |= (1 << i);
+            }
+        }
+    }
+
+    if (threading::ThreadStateInternal::getUnwindState() == UNWIND_STATE_NORMAL) {
```

Comment:
It feels weird that we combine the logic of how to process the frame with the logic of when to process the frame.  I think it might make more sense to have the caller of unwindProcessFrame know whether to actually process the frame or not, since those rules will be different based on the caller.  For example, `sys._getframe` uses this same machinery to traverse the Python stack, but that traversal will now include the rules to skip the topmost frame if we are handling a reraise.  As far as I can tell, we can never call `sys._getframe` when we are in `UNWIND_STATE_RERAISE` so the extra check *should* be ok, but it seems wrong that we check for reraises when doing `sys._getframe`.